File: Source/Game/Pawn.cpp
/* Copyright 2010 Kenneth Ferland

This file is part of Khazad.

Khazad is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Khazad is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Khazad.  If not, see <http://www.gnu.org/licenses/> */

#include <Pawn.h>

#include <DataManager.h>
#include <Game.h>
#include <Map.h>

#include <TextureManager.h>

#include <PathManager.h>
#include <PathTester.h>


Pawn::Pawn()
{
    Moving = false;
    Controller = NULL;
    CurrentMovementDirection = DIRECTION_NONE;
}

Pawn::~Pawn()
{

}

bool Pawn::Init(MapCoordinates SpawnLocation)
{
    Ogre::MaterialPtr MatPointer = TEXTURE->makeAnimatedMaterial(DATA->getLabelIndex("ANIMATION_CAT"), DATA->getLabelIndex("COLOR_BROWN"));

    Actor::Init(SpawnLocation, MatPointer, 1.0, 1.5);

    Controller = GAME->getPath()->getNewController(0, 0, LocationCoordinates);

    DestinationCoordinates = LocationCoordinates;

	Controller->setBehaviorMode(PATH_BEHAVIOR_ROUTE_TO_LOCATION);

	AnimationGroupID = DATA->getLabelIndex("ANIMATION_CAT");
	setAnimationType(DATA->getLabelIndex("ANIMATION_TYPE_IDLE"));

	//Ogre::TextureUnitState* TexUnit = Mat->getTechnique(0)->getPass(0)->getTextureUnitState(0);
	//TexUnit->setCurrentFrame(8);
	//Controller->setBehaviorMode(PATH_BEHAVIOR_WANDER_AIMLESSLY);

    return true;
}

CoolDown Pawn::AttemptMove(Direction MovementDirection)
{
    float EdgeCost = GAME->getPath()->getEdgeCost(LocationCoordinates, MovementDirection);

    if (EdgeCost != -1)
    {
        MapCoordinates NewLocation = MapCoordinates(LocationCoordinates, MovementDirection);

        Moving = true;
        float Speed = 1.0;
        MovementCoolDown = EdgeCost / (Speed / 1000.0);
        MovementStarted = TEMPORAL->getCurrentTimeTick();

        float AnimationSpeed = 0.5;
        float FramesToMove = AnimationLoopLength * AnimationSpeed * EdgeCost;
        FrameCooldown = MovementCoolDown / FramesToMove;

        // Create Vector directly from a MovementDirection?
        RenderLocationChange.x = NewLocation.X - LocationCoordinates.X;
        RenderLocationChange.y = NewLocation.Y - LocationCoordinates.Y;
        RenderLocationChange.z = NewLocation.Z - LocationCoordinates.Z;

        RenderLocationChange = RenderLocationChange / FramesToMove;
        return FrameCooldown;
        // Reduce magnitude proportional to cooldown
    } else {
    	Moving = false;
    	// Unexpected obstacle
    }
    return 1;
}

void Pawn::setAnimationType(ANIMATION_TYPE_INDEX NewAnimationType)
{
    AnimationStartIndex = CurrentFrame = DATA->getAnimationGroupData(AnimationGroupID)->getAnimationStart(NewAnimationType);
	AnimationLoopLength = DATA->getAnimationGroupData(AnimationGroupID)->getAnimationLength(NewAnimationType);

	//ActorBillboard->getMaterial()->getTechnique(0)->getPass(0)->getTextureUnitState(0)->setCurrentFrame(CurrentFrame);
}

void Pawn::AdvanceFrame()
{
    ActorBillboard->getMaterial()->getTechnique(0)->getPass(0)->getTextureUnitState(0)->setCurrentFrame(CurrentFrame++);  // advance the animation

    if (CurrentFrame == AnimationLoopLength)
    {
        CurrentFrame = AnimationStartIndex;
    }
}

CoolDown Pawn::Update()
{
    UpdateTick = TEMPORAL->getCurrentTimeTick();   // Record the current Tick


    if (Moving)
    {
        //AdvanceFrame();
        MoveRenderPosition(RenderLocationChange);

        if (UpdateTick >= (MovementStarted + MovementCoolDown))  // Done
        {
            MapCoordinates NewLocation = MapCoordinates(LocationCoordinates, CurrentMovementDirection);
            setLocation(NewLocation);
            Controller->setLocation(NewLocation);
            Moving = false;
        }
        return FrameCooldown;
    }
    else
    {
        if (LocationCoordinates == DestinationCoordinates)
        {
            if (GAME->getPath()->getDirectionFlags(LocationCoordinates) != 0) //Is the current location impassable
            {
                DestinationCoordinates = GAME->getPath()->getTester()->getRandomPassableCoordinate();  // This needs to get DIFFERENT coords each time

                while (!Controller->isDestinationReachable(DestinationCoordinates))
                {
                    DestinationCoordinates = GAME->getPath()->getTester()->getRandomPassableCoordinate();
                }
                Controller->ChangeDestination(DestinationCoordinates);
            } else {
                Controller->setBehaviorMode(PATH_BEHAVIOR_HALT);
                Moving = false;
                return 100;
            }
        }

        CurrentMovementDirection = Controller->getNextStep();

        if (CurrentMovementDirection != DIRECTION_NONE)
        {
            return AttemptMove(CurrentMovementDirection);
        }
    }

    return 1;
}

Ogre::Vector3 Pawn::getRenderPosition()
{
    return Ogre::Vector3(LocationCoordinates.X, LocationCoordinates.Y, LocationCoordinates.Z) + getRederPositionMovementAdjustment();
}

Ogre::Vector3 Pawn::getRederPositionMovementAdjustment()
{
    uint32_t MovementMultiplier = TEMPORAL->getCurrentTimeTick() - UpdateTick;
    return RenderLocationChange * MovementMultiplier;
}

File: src/Game/Pawn.java
/* Copyright 2010 Kenneth 'Impaler' Ferland

 This file is part of Khazad.

 Khazad is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Khazad is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Khazad.  If not, see <http://www.gnu.org/licenses/> */

package Game;

import Job.*;
import Core.Dice;
import Data.*;
import Data.Types.*;

import Map.Direction;
import Map.MapCoordinate;

import PathFinding.MovementModality;
import PathFinding.Navigator;
import PathFinding.PathManager;

import java.io.Serializable;

/**
 * A moving creature, derived from Actor but with the addition of a Navigator object
 * linking it to the Pathfinding engine and a set of RPG style attributes.
 *
 * @author Impaler
 */
public class Pawn extends Actor implements Serializable {

	private static final long serialVersionUID = 1;
	short CreatureTypeID;
	// Action extension of 
	long ActionDuration = 1;
	long ActionStarted;
	// Movement and Pathfinding
	Navigator PathNavigator;
	boolean Moving;
	Direction CurrentMovementDirection;
	// Job/Task Managment
	private Task CurrentTask;
	public Job PrimaryJob;
	public Job BreakJob;
	public boolean onBreak;
	// The 14 Basic Attributes
	Dice AttributeDice;
	byte[] BasicAttributes;
	private String firstName;
	private String lastName;
	private Gender gender;

	public Pawn(short CreatureTypeID, int id, int Seed, MapCoordinate SpawnLocation) {
		super(id, SpawnLocation);

		this.CreatureTypeID = CreatureTypeID;
		Moving = false;
		PathNavigator = new Navigator(SpawnLocation, new MovementModality(MovementModality.MovementType.MOVEMENT_TYPE_WALK, 1, 1));
		CurrentMovementDirection = Direction.DIRECTION_NONE;

		AttributeDice = new Dice();
		AttributeDice.seed(id ^ Seed);

		DataManager Data = DataManager.getDataManager();
		CreatureData CreatureDataEntry = Data.getCreatureData(CreatureTypeID);
		CreatureSizeData CreatureSizeDataEntry = Data.getCreatureSizeData(CreatureDataEntry.SizeID);

		BasicAttributes = new byte[Data.getNumBaseAttributes()];
		for (int i = 0; i < BasicAttributes.length; i++) {
			BasicAttributes[i] = (byte) (AttributeDice.roll(1, 4) + AttributeDice.roll(1, 4));
			BasicAttributes[i] += CreatureDataEntry.AttributeModifierVales[i];  //Size class adjustment
			BasicAttributes[i] += CreatureSizeDataEntry.AttributeModifierVales[i];  //Size class adjustment
		}
		if (AttributeDice.roll(1,3) == 1) {
			gender = Gender.GENDER_FEMALE;
		} else {
			gender = Gender.GENDER_MALE;
		}
	}

	public Navigator getNavigator() {
		return PathNavigator;
	}

	public long attemptMove(Direction MovementDirection) {
		float EdgeCost = PathManager.getSingleton().getEdgeCost(LocationCoordinates, MovementDirection, PathNavigator.getMovementModality());
		final int speedIndex = DataManager.getDataManager().getLabelIndex("BASIC_ATTRIBUTE_SPEED");
		if (EdgeCost != -1) {
			return (int) (EdgeCost / ((float) BasicAttributes[speedIndex] / 7.0) * TICKS_PER_SECOND);
		} else {
			return 1;  // signal falure to job manager?
		}
	}

	public long updatePosition() {
		setLocation(new MapCoordinate(LocationCoordinates, CurrentMovementDirection));
		CurrentMovementDirection = PathNavigator.getNextStep();

		if (CurrentMovementDirection == Direction.DIRECTION_DESTINATION) {
			CurrentTask.Completed = true;
			return 1;
		} else {
			return attemptMove(CurrentMovementDirection);
		}
	}

	public boolean isDestinationReachable(MapCoordinate Destination) {
		return PathNavigator.isDestinationReachable(Destination);
	}

	@Override
	public final void setLocation(MapCoordinate NewLocation) {
		super.setLocation(NewLocation);
		PathNavigator.setLocation(NewLocation);
	}

	public Task getTask() {
		return CurrentTask;
	}

	public void setTask(Task NewTask) {
		if (CurrentTask != NewTask) {
			CurrentTask = NewTask;
		}
	}

	public float getActionFraction(long CurrentTick) {
		return (CurrentTick - ActionStarted) / ((float) ActionDuration);
	}

	public void setMovementDiretion(Direction newDirection) {
		CurrentMovementDirection = newDirection;
	}

	public Direction getMovementDirection() {
		return CurrentMovementDirection;
	}

	public MovementModality getMovementModality() {
		return PathNavigator.getMovementModality();
	}

	public byte getBasicAttribute(short AttributeID) {
		return BasicAttributes[AttributeID];
	}

	public Task findTask() {
		PrimaryJob = new WanderJob();
		CurrentTask = PrimaryJob.nextTask(this);
		return CurrentTask;
	}

	@Override
	long wake(long CurrentTick) {
		//super.Wake(CurrentTick);
		if (CurrentTask != null) {
			if (CurrentTask.Completed) {
				CurrentTask.finalizeTask(CurrentTick, this);
				// check for other needs, leave current Job if nessary
				setTask(CurrentTask.ParentJob.nextTask(this));
			}

			if (!CurrentTask.Begun) {
				ActionDuration = CurrentTask.beginTask(this);
			} else {
				ActionDuration = CurrentTask.continueTask(this);
				if (CurrentTask.Completed) {
					CurrentTask.finalizeTask(CurrentTick, this);
					setTask(CurrentTask.ParentJob.nextTask(this));
				}
			}
			ActionStarted = CurrentTick;
			WakeTick = CurrentTick + ActionDuration;
		} else {
			WakeTick = CurrentTick + 1;
		}

		return WakeTick;
	}

	public String getFirstName() {
		return firstName;
	}

	public void setFirstName(String firstName) {
		this.firstName = firstName;
	}

	public String getLastName() {
		return lastName;
	}

	public void setLastName(String lastName) {
		this.lastName = lastName;
	}
	
	public String getName() {
		return this.firstName + ' ' + this.lastName;
	}

	public Gender getGender() {
		return gender;
	}

	public void setGender(Gender gender) {
		this.gender = gender;
	}

	public String getGenderText() {
		if (gender == Gender.GENDER_MALE) {
			return "Male";
		} else if (gender == Gender.GENDER_FEMALE) {
			return "Female";
		}
		return "Unknown";
	}
}
